lib/SILOptimizer/Utils/Existential.cpp

```diff
@@ -309,3 +310,73 @@ ConcreteExistentialInfo::ConcreteExistentialInfo(Operand &openedUse) {
   }
   assert(isValid());
 }
+
+/// Initialize a ConcreteExistentialInfo based on the already computed concrete
+/// type and protocol declaration. It determines the OpenedArchetypeDef
+/// and SubstituionMap for the ArgOperand argument.
+/// We need the OpenedArchetypeDef to be able to cast it to the concrete type.
+/// findInitExistential helps us determine this OpenedArchetypeDef. For cases
+/// where OpenedArchetypeDef can not be found from findInitExistential (because
+/// there was no InitExistential), then we do
+/// extra effort in trying to find an OpenedArchetypeDef for AllocStackInst
+/// using getAddressOfStackInit.
+ConcreteExistentialInfo::ConcreteExistentialInfo(Operand &ArgOperand,
+                                                 const CanType &ConcreteTy,
+                                                 ProtocolDecl *Protocol)
+    : ConcreteExistentialInfo(ArgOperand) {
+
+  // If we found an InitExistential, assert that ConcreteType we determined is
+  // same as ConcreteTy argument.
+  if (InitExistential) {
+    assert(ConcreteType == ConcreteTy);
+    assert(isValid());
+    return;
+  }
+
+  ConcreteType = ConcreteTy;
+  OpenedArchetypeDef = ArgOperand.get();
+
+  // If findInitExistential call from the other constructor did not update the
+  // OpenedArchetypeDef (because it did not find an InitExistential) and that
+  // the original Arg is an alloc_stack instruction, then we determine the
+  // OpenedArchetypeDef using getAddressOfStackInit. Please keep in mind that an
+  // alloc_stack can be an argument to apply (which could have no associated
+  // InitExistential), thus we need to determine the OpenedArchetypeDef for it.
+  // This is the extra effort.
+  SILInstruction *User = ArgOperand.getUser();
+  SILModule &M = User->getModule();
+  if (auto *ASI = dyn_cast<AllocStackInst>(OpenedArchetypeDef)) {
+    bool copied = false;
+    if (SILValue Src = getAddressOfStackInit(ASI, User, copied)) {
+      OpenedArchetypeDef = Src;
+    }
+    isCopied = copied;
+  }
+
+  // Bail, if we did not find an opened existential.
+  if (!(isa<OpenExistentialRefInst>(OpenedArchetypeDef) ||
+        isa<OpenExistentialAddrInst>(OpenedArchetypeDef)))
+    return;
+
+  // We have the open_existential; we still need the conformance.
+  auto ConformanceRef =
+      M.getSwiftModule()->lookupConformance(ConcreteType, Protocol);
```

Comment:
I'm still curious what @slavapestov has to say about calling `lookupConformance` here. It seems like we should come up with some rules for when it's ok for the optimizer to do this and enforce it in the API.

Reply:
In this case it's fine to call lookupConformance() -- there's no way for the type checker to anticipate a conformance is needed here and store it somewhere. However, the resulting conformance might be incomplete, which will cause problems later in IRGen if you try to use it.

So I would add a check for the conformance being complete.

Also please try to come up with a test case that would fail without this check. It will require the conforming type to be defined in another file, so that when you build in non-WMO mode, the type checker will not check the conformance.

Reply:
... OK, but since this optimization only applies in WMO mode, you won't be able to get a test case. In that case, please just assert that the conformance is complete.